pgml-apps/discord-bot/bot.py

```diff
@@ -0,0 +1,153 @@
+    def ingest(self, path: str, collection_name: str):
+        docs = self.load_documents(path)
+        content = self.create_content(docs)
+        collection = self.pgml.create_or_get_collection(collection_name)
+        collection.upsert_documents(content)
+        embedding_model_id = self.register_embedding_model(collection)
+        splitter_id = self.register_text_splitter(collection)
+        collection.generate_chunks(splitter_id=splitter_id)
+        collection.generate_embeddings(model_id=embedding_model_id, splitter_id=splitter_id)
+    
+    def create_or_get_collection(self, collection_name: str):
+        return self.pgml.create_or_get_collection(collection_name)
+
+    # Loads markdown documents from a directory
+    def load_documents(self, path: str):
+        print(f"Loading documents from {path}")
+        loader = DirectoryLoader(path, glob='*.md')
+        docs = loader.load()
+        print(f"Loaded {len(docs)} documents")
+        return docs
+
+    # Prepare content by iterating over each document
+    def create_content(self, docs):
+        return [{"text": doc.page_content, "source": doc.metadata['source']} for doc in docs]
+
+
+    # Register an embedding model to the collection
+    def register_embedding_model(self, collection):
+        embedding_model_id = collection.register_model(
+            model_name="hkunlp/instructor-xl",
+            model_params={"instruction": "Represent the document for retrieval: "},
+        )
+        return embedding_model_id
+
+    # Register a text splitter to the collection
+    def register_text_splitter(self, collection, chunk_size: int = 1500, chunk_overlap: int = 40):
+        splitter_id = collection.register_text_splitter(
+            splitter_name="RecursiveCharacterTextSplitter",
+            splitter_params={"chunk_size": chunk_size, "chunk_overlap": chunk_overlap},
+        )
+        return splitter_id
+
+    # Run an SQL query and return the result
+    async def run_query(self, statement: str, sql_params: tuple = None):
+        conn = self.pool.getconn()
+        cur = conn.cursor()
+        try:
+            cur.execute(statement, sql_params)
+            return cur.fetchone()
+        except Exception as e:
+            print(e)
+        finally:
+            cur.close()
+
+    # Query a collection with a string and return vector search results
+    async def query_collection(self, collection_name: str, query: str):
+        collection = self.pgml.create_or_get_collection(collection_name)
+        return collection.vector_search(
+            query,
+            top_k=3,
+            model_id=2,
+            splitter_id=2,
+            query_parameters={"instruction": "Represent the question for retrieving supporting documents: "},
+        )
+
+    # Start the Discord bot, listen to messages in 'bot-testing' channel and handle the messages
+    def start(self, collection_name: str, discord_token: str, channel_name: str):
```

Comment:
Just general code organization, I'd put a method like this that is not async, at the top of the file. Are these otherwise alphabetical?